crates/transpiler/src/passes/basis_translator/compose_transforms.rs

```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
 };
 use smallvec::SmallVec;
 
-use crate::equivalence::CircuitFromPython;
+use super::errors::BasisTranslatorError;
 
 // Custom types
 pub type GateIdentifier = (String, u32);
 pub type BasisTransformIn = (SmallVec<[Param; 3]>, CircuitFromPython);
 pub type BasisTransformOut = (SmallVec<[Param; 3]>, DAGCircuit);
 
+static STD_GATE_MAPPING: OnceLock<HashMap<&str, StandardGate>> = OnceLock::new();
+static STD_INST_MAPPING: OnceLock<HashSet<&str>> = OnceLock::new();
+
 pub(super) fn compose_transforms<'a>(
-    py: Python,
     basis_transforms: &'a [(GateIdentifier, BasisTransformIn)],
     source_basis: &'a IndexSet<GateIdentifier, ahash::RandomState>,
     source_dag: &'a DAGCircuit,
-) -> PyResult<IndexMap<GateIdentifier, BasisTransformOut, ahash::RandomState>> {
+) -> Result<IndexMap<GateIdentifier, BasisTransformOut, ahash::RandomState>, BasisTranslatorError> {
     let mut gate_param_counts: IndexMap<GateIdentifier, usize, ahash::RandomState> =
         IndexMap::default();
-    get_gates_num_params(source_dag, &mut gate_param_counts)?;
+    get_gates_num_params(source_dag, &mut gate_param_counts);
     let mut mapped_instructions: IndexMap<GateIdentifier, BasisTransformOut, ahash::RandomState> =
         IndexMap::with_hasher(ahash::RandomState::default());
 
     for (gate_name, gate_num_qubits) in source_basis.iter().cloned() {
         let num_params = gate_param_counts[&(gate_name.clone(), gate_num_qubits)];
 
-        // TODO: Follow-up by replacing with Param::ParamterExpression(Symbol),
-        // as ParameterVector is not exposed in Rust. Maybe we ought to add a function
-        // for convenient construction, if it would be useful.
-        let placeholder_params: SmallVec<[Param; 3]> = PARAMETER_VECTOR
-            .get_bound(py)
-            .call1((&gate_name, num_params))?
-            .extract()?;
+        // The last usage of Python left is the parameter vector here.
+        let mut placeholder_params: SmallVec<[Param; 3]> = (0..num_params as u32)
+            .map(|idx| {
+                Param::ParameterExpression(
+                    ParameterExpression::from_symbol(Symbol::new(
+                        &format!("{gate_name}-{idx}"),
+                        None,
+                        None,
```

Comment:
We could maybe use the native indexing support here, the symbols have the feature for it
```suggestion
                    ParameterExpression::from_symbol(Symbol::new(
                        gate_name,
                        None,
                        Some(idx),
```

Reply:
I had it like this before, but was suggested not to use it since there is no `ParameterVector` linked here.

Reply:
Hmm do we need a linked parameter vector? From the last discussion in the dev meeting we said that the linked vector wouldn't be functionally necessary anywhere, right? 🤔

Reply:
Added in cbe4e1d868234c50cce87a47723cbac688aea0e7